r_code/placebo_creation.R

```diff
@@ -0,0 +1,217 @@
+                              treat_indicator = "treatperiod_0",
+                              outcome_var = "target", 
+                              counterfac_var = "counter_factual") {
+  # Generates a placebo tibble, using matching methods to select
+  # placebo-treated entries as those most similar to truly-treated.
+  
+  # Args
+  # data_full: long-form dataframe with both treated and control entries.
+  #            rows of the df represent period-entry combinations 
+  #            (eg N (total num of entry) rows for period t).
+  #            each row should have a treatment indicator (treat_indicator), 
+  #            a period number (time_var),
+  #            an individual ID (id_var), and an outcome (outcome_var)
+  #            for associated with that period-ID combination
+  # id_var: column name of numeric, unique ID representing the entry (unit)
+  # time_var:column name of numeric period number indicating the time period, 
+  #          in increasing order (eg 0 is the first time, 120 is the last)
+  # treat_indicator:column name of binary (0, 1) indicator for whether id_var i 
+  #                 in time_var t was treated. 
+  #                 Once treated, must always be treated.
+  
+  
+  # Output
+  # placebo_df_long, a tibble of the same format as data_full, entirely
+  # consisting of donor units, some of which are placebo-treated 
+  # (based on matching).
+  
+  # Split the dataset based on whether the unit is ever treated
+  tr_entries <- data_full %>%
+    dplyr::filter(!!as.name(treat_indicator) > 0) %>%
+    dplyr::distinct(!!as.name(id_var)) %>%
+    dplyr::pull()
+  
+  ct_entries <- setdiff(data_full %>%
+                          dplyr::distinct(!!as.name(id_var)) %>%
+                          dplyr::pull(), tr_entries)
+  # Create a dataframe pf the subset of control units
+  cd <- data_full %>% dplyr::filter(!!as.name(id_var) %in% ct_entries)
+  cd <- merge(
+    cd,
+    data.frame(
+      entry = ct_entries,
+      rank = seq_along(ct_entries)
+    )
+  )
+  
+  # Pivot the long cd dataframe to wide. Each row will represent an id_var, 
+  #with columns for the outcome_var at each time period
+  cd_for_match <- tidyr::pivot_wider(
+    data = cd %>%
+      dplyr::arrange(!!as.name(time_var), !!as.name(id_var)),
+    names_from = !!as.name(time_var),
+    id_cols = c(!!as.name(id_var)),
+    values_from = c(!!as.name(outcome_var))
+  )
+  
+  # Store treated data in a tibble
+  treated_to_match <- data_full %>%
+    dplyr::filter(!!as.name(id_var) %in% tr_entries)
+  
+  #Create a variable indicating the time treatment is assigned by unit
+  treated_to_match <- treated_to_match %>%
+    dplyr::group_by(!!as.name(id_var)) %>%
+    dplyr::mutate(
+      Treatment_Period =
+        length(!!as.name(treat_indicator)) - sum(!!as.name(treat_indicator)) + 1
+    ) %>%
+    dplyr::ungroup()
+  
+  # for pivoting, potential issues arise if we have several time varying 
+  # covariates -- added as TODO.
```

Comment:
I see that the tidyverse documentation does not specify a todo syntax even though it should.  Let's use the Python style guide syntax (https://google.github.io/styleguide/pyguide.html#312-todo-comments):

```
TODO(alexdkellogg): Revisit whether pivoting will cause issues in the presence of several time varying covariates.
```

There are at least two advantages:
1) The syntax is regular so it can be easily searched.
2) The scope of the todo is easier to discern since it starts with the syntax and ends with a period after one or more sentences.